trackintel/preprocessing/triplegs.py

```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
-    tpls = tpls_input.copy()
-    stps = stps_input.copy()
+    assert "activity" in spts.columns, "staypoints need the column 'activities' to be able to generate trips"
 
-    # if the triplegs already have a column "trip_id", we drop it
+    # Copy the input because we add a temporary columns
+    tpls = tpls.copy()
+    spts = spts.copy()
+    gap_threshold = pd.to_timedelta(gap_threshold, unit="min")
+
+    # If the triplegs already have a column "trip_id", we drop it
     if "trip_id" in tpls:
         tpls.drop(columns="trip_id", inplace=True)
+        warnings.warn("Deleted existing column 'trip_id' from tpls.")
 
-    # if the staypoints already have any of the columns  "trip_id", "prev_trip_id", "next_trip_id", we drop them
+    # if the staypoints already have any of the columns "trip_id", "prev_trip_id", "next_trip_id", we drop them
     for col in ["trip_id", "prev_trip_id", "next_trip_id"]:
-        if col in stps:
-            stps.drop(columns=col, inplace=True)
+        if col in spts:
+            spts.drop(columns=col, inplace=True)
+            warnings.warn(f"Deleted column '{col}' from spts.")
 
     tpls["type"] = "tripleg"
-    stps["type"] = "staypoint"
+    spts["type"] = "staypoint"
 
     # create table with relevant information from triplegs and staypoints.
-    stps_tpls = stps[["started_at", "finished_at", "user_id", "type", "activity"]].append(
-        tpls[["started_at", "finished_at", "user_id", "type"]]
+    spts_tpls = pd.concat(
+        [
+            spts[["started_at", "finished_at", "user_id", "type", "activity"]],
+            tpls[["started_at", "finished_at", "user_id", "type"]],
+        ]
     )
+    # transform nan to bool
+    spts_tpls["activity"].fillna(False, inplace=True)
 
     # create ID field from index
-    stps_tpls["id"] = stps_tpls.index
+    spts_tpls["id"] = spts_tpls.index
+
+    spts_tpls.sort_values(by=["user_id", "started_at"], inplace=True)
+    spts_tpls.reset_index(inplace=True, drop=True)
+
+    # delete intermediate activities
+    # intermediate activities are activities that are surrounded by other activities
+    # therefore they are neither at the end, at the start, nor within a trip
+    first_activities, inter_activities, last_activities = _get_activity_masks(spts_tpls)
+    spts_tpls["first_activity"] = first_activities
+    spts_tpls["last_activity"] = last_activities  # trip may start here
+    spts_tpls = spts_tpls[~inter_activities]  # no trip interaction
+    spts_tpls["started_at_next"] = spts_tpls["started_at"].shift(-1)
+
+    # conditions for new trip
+    # start new trip if the user changes
+    condition_new_user = spts_tpls["user_id"] != spts_tpls["user_id"].shift(1)
+
+    # start new trip if there is a new activity
+    condition_new_activity = spts_tpls["last_activity"]
+
+    # gap conditions
+    # start new trip after a gap
+    gap = (spts_tpls["started_at_next"] - spts_tpls["finished_at"]) > gap_threshold
```

Comment:
Current `spts_tpls` is after removing the `inter_activity` (line 133). Isn't that the spts are now filtered, and spt information might be lost?

Reply:
No It should not, as no trip can start or end on these intermediate activities.
But I think I do not fully understand your question.

Reply:
I cleaned up the code a bit there now it isn't necessary anymore to filter for `inter_activities`.